lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp

```diff
@@ -196,19 +212,32 @@ class LowerXMRPass : public LowerXMRBase<LowerXMRPass> {
         return inst.emitOpError("cannot lower ext modules with RefType ports");
       // Reference ports must be removed.
       refPortsToRemoveMap[inst].push_back(portNum);
-      // Drop dead instance ports.
+      // Drop the dead-instance-ports.
       if (instanceResult.use_empty())
         continue;
       auto refModuleArg = refMod.getArgument(portNum);
-      // Get the remote RefSendOp, that flows through the module ports.
-      auto remoteOpPath = getRemoteRefSend(refModuleArg);
-      if (!remoteOpPath)
-        return failure();
-
-      auto pathToRefSend = remoteOpPath.getValue().vec();
-      pathToRefSend.push_back(getInnerRefTo(inst));
-      reachingRefSendAt[instanceResult] =
-          ArrayAttr::get(inst.getContext(), pathToRefSend);
+      if (inst.getPortDirection(portNum) == Direction::Out) {
+        // For output instance ports, the dataflow is into this module.
+        // Get the remote RefSendOp, that flows through the module ports.
+        // If dataflow at remote module argument does not exist, error out.
+        auto remoteOpPath = getRemoteRefSend(refModuleArg);
+        if (!remoteOpPath)
+          return failure();
+        // Get the path to reaching refSend at the referenced module argument.
+        // Now append this instance to the path to the reaching refSend.
+        addReachingSendsEntry(instanceResult, getInnerRefTo(inst),
+                              remoteOpPath);
+      } else {
+        // For input instance ports, the dataflow is into the referenced module.
+        // Input RefType port implies, generating an upward scoped XMR.
+        // No need to add the instance context, since downward reference must be
+        // through single instantiated modules.
+        if (multiplyInstantiated)
+          return refMod.emitOpError(
+                     "multiply instantiated module with input RefType port '")
+                 << portNum << "'";
```

Comment:
Nit/suggestion: Would it be better to print port name here (like in the other port error message)?